### PR TITLE
Test framework/guide

### PR DESCRIPTION
**Pull request by nhtruong**
- Reduced noise when running spec tests
- `--verbose` now provides better context for non-2XX responses.
- Moved `linter/utils` to `_utils` because it's also being used by the tester tool (To avoid misleading stacktrace when an error appears in utils)
- Added SPEC_TESTING.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: SPEC_TESTING.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+* [SPEC TESTING](#spec-testing)
+  * [Running Spec Tests Locally](#running-spec-tests-locally)
+  * [Writing Spec Tests](#writing-spec-tests)
+    * [Simple Test Story](#simple-test-story)
+    * [Using Output from Previous Chapters](#using-output-from-previous-chapters)
+<!-- TOC -->
+
+# SPEC TESTING
+
+We devise our own test framework to test the spec against an OpenSearch cluster. We're still adding more features to the framework as the needs arise, and this document will be updated accordingly. This test framework has also been integrated into the repo's CI/CD pipeline. Checkout the [test-spec](.github/workflows/test-spec.yml) workflow for more details.
```
Comment (Xtansia) [inline SPEC_TESTING.md:11]: I think tense wise this should be `We have devised...`

File: tools/src/linter/InlineObjectSchemaValidator.ts [hunk 2]
```diff
@@ -10,8 +10,8 @@
 import type NamespacesFolder from './components/NamespacesFolder'
 import type SchemasFolder from './components/SchemasFolder'
 import { type ValidationError } from 'types'
-import { SchemaVisitor } from './utils/SpecificationVisitor'
-import { is_ref, type MaybeRef, SpecificationContext } from './utils'
+import { SchemaVisitor } from '../_utils/SpecificationVisitor'
+import { is_ref, type MaybeRef, SpecificationContext } from '../_utils'
```
Comment (Xtansia) [inline tools/src/linter/InlineObjectSchemaValidator.ts:14]: What's the reason for the leading underscore on `_utils`?

File: tools/src/tester/test.ts [hunk 3]
```diff
@@ -7,22 +7,23 @@
...
+  OPENSEARCH_USERNAME_OPTION,
+  OpenSearchHttpClient
 } from '../OpenSearchHttpClient'
 import ChapterReader from './ChapterReader'
 import ChapterEvaluator from './ChapterEvaluator'
 import OperationLocator from './OperationLocator'
 import SchemaValidator from './SchemaValidator'
 import StoryEvaluator from './StoryEvaluator'
-import { ConsoleResultLogger } from './ResultLogger'
+import {ConsoleResultLogger} from './ResultLogger'
```
Comment (Xtansia) [inline tools/src/tester/test.ts:26]: I'm surprised the linter isn't complaining about this as it should be enforcing spacing around the braces (or at least it was)

**Review by Xtansia — CHANGES_REQUESTED** [review]

File: tools/src/tester/test.ts (hunk 3 above)
Comment (nhtruong) [reply tools/src/tester/test.ts:26]: I setup my IDE to run `--fix` on save. I thought the linter wanted space too.

File: tools/src/linter/InlineObjectSchemaValidator.ts (hunk 2 above)
Comment (nhtruong) [reply tools/src/linter/InlineObjectSchemaValidator.ts:14]: To differentiate it from other folders like `tester` `linter` `merger` where they are tools while `_utils` is not a tool.

File: SPECIFICATION_TESTING.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
+<!-- TOC -->
```
Comment (dblock) [inline SPECIFICATION_TESTING.md:1]: We don't have these markers in other docs, maybe remove? The [Markdown all in one](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one) vscode plugin doesn't require them.

File: tools/src/tester/test.ts (hunk 3 above)
Comment (dblock) [reply tools/src/tester/test.ts:26]: This looks suspicious. Other files have a space. Something is off with the eslint rule file being used?

https://github.com/opensearch-project/opensearch-api-specification/blob/main/tools/src/coverage/coverage.ts#L12

**Review by dblock — CHANGES_REQUESTED** [review]
Nits + let's resolve this linter problem.

Feel free to break up the PR, the doc looks good.

File: SPECIFICATION_TESTING.md (hunk 4 above)
Comment (nhtruong) [reply SPECIFICATION_TESTING.md:1]: It appears in [DEVELOPER_GUIDE](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md?plain=1#L1), too. Shall I remove it there too?

Comment (dblock) [reply SPECIFICATION_TESTING.md:1]: Yes, but I am nitpicking.

File: tools/src/tester/test.ts (hunk 3 above)
Comment (nhtruong) [reply tools/src/tester/test.ts]: Fixed it. Didn't dig too deep but `object-curly-spacing` rule was deprecated in ESLint v8.53.0 and moved to a different plugin. Installing the plugin and enabled that rule solved it.

File: SPECIFICATION_TESTING.md (hunk 4 above)
Comment (dblock) [reply SPECIFICATION_TESTING.md:1]: Note that it's not valid markdown ;)

**Review by dblock — APPROVED** [review]

File: SPECIFICATION_TESTING.md (hunk 4 above)
Comment (Xtansia) [reply SPECIFICATION_TESTING.md:1]: It's just an HTML comment, and inline raw HTML is valid according to:
- [GitHub Flavored Markdown](https://github.github.com/gfm/#raw-html)
- [CommonMark](https://spec.commonmark.org/0.31.2/#raw-html)
- [John Gruber's Original Definition](https://daringfireball.net/projects/markdown/syntax#html)

**Review by Xtansia — APPROVED** [review]

File: SPECIFICATION_TESTING.md (hunk 4 above)
Comment (dblock) [reply SPECIFICATION_TESTING.md:1]: I stay corrected.

**Pull request merged**